enclave-runner/src/usercalls/mod.rs

```diff
@@ -1340,17 +1334,29 @@ impl<'tcs> IOHandlerInput<'tcs> {
         }
 
         if ret.is_none() {
+            let start = Instant::now();
             loop {
-                let ev = if wait {
-                    Ok(self.tcs.event_queue.next().await.unwrap())
-                } else {
-                    match self.tcs.event_queue.try_next() {
+                let ev = match timeout {
+                    WAIT_INDEFINITE => self.tcs.event_queue.next().await.ok_or(()),
+                    WAIT_NO => match self.tcs.event_queue.try_next() {
                         Ok(Some(ev)) => Ok(ev),
-                        Err(_) => break,
                         Ok(None) => Err(()),
+                        Err(_) => break,
+                    },
+                    timeout => {
+                        let remaining = match Duration::from_nanos(timeout).checked_sub(start.elapsed()) {
+                            None => break,
+                            Some(ref duration) if duration.as_nanos() == 0 => break,
```

Comment:
I'm not sure how likely this is, but it's possible we loop through multiple times and receive unrelated events. In case we have used all of our timeout allowance, i.e. when remaining goes negative (`None`) or zero (`Some(duration) if ...`) we would want to break the loop.

Reply:
Yeah I understand that, I'm just saying that statistically, the result will always either be `None` or the timeout will not be expired yet since we're dealing with nanosecond precision

Reply:
as discussed out of band, experiment shows that this can be likely for certain values of timeout.